datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaProducerWrapper.java

```diff
@@ -177,7 +182,12 @@ void assignTask(DatastreamTask task) {
   }
 
   void unassignTask(DatastreamTask task) {
-    _tasks.remove(task);
+    boolean taskPresent = _tasks.remove(task);
+    // whenever a task is unassigned the kafka producer should be shutdown to ensure that
+    // there are no in-flight sends. Any further send will not work.
+    if (taskPresent) {
+      shutdownProducer();
```

Comment:
One thing to note here is that since shutdownProducer() requires obtaining the lock, this operation can get stuck waiting on this lock. Operations such as flush can take a while (up to 15 minutes for BMM with the current flush timeout). Once concern might be that unassignTask() is called from the onAssignmentChange() threads, right? And have a limited time of 30-60 seconds within which it needs to complete?

Reply:
handled it.